### DISCUSSION TITLE
Understanding MCTC line type

### DISCUSSION THREAD

**Discussion by malloryprow** [dtcenter/METplus#1916] [Statistical Computation] [answered]
Hello,

I set up a ptype use case that outputs CTC (set it up based on https://metplus.readthedocs.io/en/latest/generated/model_applications/precipitation/PointStat_fcstMULTI_obsMETAR_PtypeComparisons.html#sphx-glr-generated-model-applications-precipitation-pointstat-fcstmulti-obsmetar-ptypecomparisons-py).

For EVS, for ptype we also want MCTC. I'm having some trouble understanding that output from this line type, but also maybe is has to do with me set up?

Here is how I have it set up:
- Model data: Take the CRAIN, CSNOW, CFRZR, and CICEP variables from the model to "blend" into one variable "PTYPE". Where CRAIN = 1, value assigned is 1. Where CSNOW = 1, value assigned is 2. Where CFRZR = 1, values assigned is 3. Where CICEP = 1, values assigned is 4.
- Observation data: Use "PRWE" from the prebufr files and convert using pb2nc. 

So in my METplus conf that runs point_stat:
- Use censor_val and censor_thresh to match up the observation data values to the model values (OBS_VAR1_OPTIONS = censor_thresh = [>=161&&<=163,>=171&&<=173, >=164&&<=166, >=174&&<=176]; censor_val=[1.0, 2.0, 3.0, 4.0];).
- I have set: FCST_VAR1_THRESH = ge1.0, ge2.0, ge3.0, ge4.0 and OBS_VAR1_THRESH = ge1.0, ge2.0, ge3.0, ge4.0

Here is an example output I am getting:

> MCTC      292 5 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 0 32 18 0 242 0.2

Looking at the MCTC line type out in the the docs: https://met.readthedocs.io/en/latest/Users_Guide/point-stat.html#id12. That first number I understand is the match pairs. Next is "N_CAT" which has a value of 5. <b>Why does this have a value of 5 when 4 thresholds were given?</b>

Then we would have

- F1_O1 = 0
- F1_O2 = 0
- F1_O3 =  0
- F1_O4 = 0
- F1_O5  = 0
- F2_O1 =  0
- F2_O2 =  0
- F2_O3 =  0
- F2_O4 =  0
- F2_O5 =  0
- F3_O1 =  0
- F3_O2 =  0
- F3_O3 =  0
- F3_O4 =  0
- F3_O5 =  0
- F4_O1 =  0
- F4_O2 =  0
- F4_O3 =  0
- F4_O4 =  0
- F4_O5 =  0
- F5_O1 =  0
- F5_O2 =  32
- F5_O3 =  18
-  F5_O4 = 0
- F5_O5 =  242
- EC_VALUE = 0.2

<b>How do these forecast and observation categories line up with the thresholds?</b>

**Comment by malloryprow** [comment]
I did find a bug in my ptype blending code, but the questions still remain!

**Comment by JohnHalleyGotway** [comment]
@malloryprow thanks for describing your approach. That's very interesting and makes a lot of sense.

It does strike me that there's likely a category missing for the null case, where no precipitation is occurring. I imagine that'd often be a large percentage of the grid. You could considering encoding the "no-precip" case as a value of 0, in both the forecast and observation datasets.

In addition, you might consider what should happen when a grid point has more than one precipitation type occurring at the same time. I was actually involved in a very similar project a couple years back comparing model precipitation type to mping point observations. And the scientists, including @jwolff-ncar, spent time carefully sifting through those questions and details.

@jwolff-ncar are you able to share the final report of that comparison with @malloryprow? I recall that the report included details about the methodology we chose.

Regarding your question about the number of thresholds, consider the simple 2x2 case which produces the CTC output line type from MET. The user chooses 1 thresholds and that defines 2 categories: either the threshold condition is met or it is not = 2 categories. In general in MET, defining `n` thresholds produces `n+1` categories. In your case, `cat_thresh = [ ge1.0, ge2.0, ge3.0, ge4.0 ];` produces 5 categories, the 5-th one being `<1.0`.

I'd recommend running `plot_data_plane` on your merged forecast input field to make sure the merging is happening the way you expect. I'd also recommend writing out the MPR line type while you're still in development MODE. That'll enable you see that the actual FCST and OBS values being compared after the censoring logic has been applied.

**Comment by jwolff-ncar** [comment]
Report emailed!

**Comment by malloryprow** [comment, marked as answer]
Thanks @JohnHalleyGotway and @jwolff-ncar.

So would that F1_O1 column be F(>=1.0)_O(>=1.0)? We can't use == with MCTC can we?

I think what we would be looking for where counts where fcst_rain&obs_rain, fcst_rain&obs_snow, etc. Using the threshold ge1 would include all the ptypes would seem kind of useless.

**Reply by JohnHalleyGotway** [reply]
@malloryprow, no, by convention, strict equality is not allowed in MCTC computation because, logically speaking, it would never include all possible values of a continuous variable. We could end up with values that don't fall in any category. When trying to run that way, this is the error message you'd get:
```
ERROR  : check_mctc_thresh() -> when verifying using multi-category contingency tables, the thresholds must be monotonically increasing and be of the same inequality type (lt, le, gt, or ge).
```
We could, of course, modify the "rules" in whatever way we'd like, but I worry that may yield more unexpected, confusing results than how we have it now.

By way of example, I ran point_stat on relative humidity data using censoring options to bucket it into categories:
```
   field = [
      {
        name          = "RH";
        level         = [ "Z2" ];
        censor_thresh = [ <60, >=60&&<70, >=70&&<80, >=80&&<90, >=90 ];
        censor_val    = [ 0, 1, 2, 3, 4 ];
        cat_thresh    = [ ge1.0, ge2.0, ge3.0, ge4.0 ];
      }
   ];
```
And that creates this 5x5 MCTC table:
```
VERSION MODEL DESC FCST_LEAD FCST_VALID_BEG  FCST_VALID_END  OBS_LEAD OBS_VALID_BEG   OBS_VALID_END   FCST_VAR FCST_UNITS FCST_LEV OBS_VAR OBS_UNITS OBS_LEV OBTYPE VX_MASK INTERP_MTHD INTERP_PNTS FCST_THRESH             OBS_THRESH              COV_THRESH ALPHA LINE_TYPE TOTAL N_CAT F1_O1 F1_O2 F1_O3 F1_O4 F1_O5 F2_O1 F2_O2 F2_O3 F2_O4 F2_O5 F3_O1 F3_O2 F3_O3 F3_O4 F3_O5 F4_O1 F4_O2 F4_O3 F4_O4 F4_O5 F5_O1 F5_O2 F5_O3 F5_O4 F5_O5 EC_VALUE
V11.0.0 WRF   NA   360000    20070331_120000 20070331_120000 000000   20070331_103000 20070331_133000 RH       %          Z2       RH      %         Z2      ADPSFC FULL    NEAREST     1           >=1.0,>=2.0,>=3.0,>=4.0 >=1.0,>=2.0,>=3.0,>=4.0 NA         NA    MCTC       3967     5   320    77    62    54    28   167    83    79    69    28    67    63   118   156   106    39    76   138   366   275    16    27   108   515   930      0.2
```
The 4 thresholds listed in the `FCST_THRESH = >=1.0,>=2.0,>=3.0,>=4.0` column define 5 categories in MET which span the number line. They are defined as:
Cat 1: <1
Cat 2: >=1 && <2
Cat 3: >=2 && <3
Cat 4: >=3 && <4
Cat 5: >=4

Those 5 categories fully cover all numeric possibilities. When you define N thresholds, the N+1 threshold is implicitly defined based on the inequality type used.

Checking that with this sample data, I see that `sum(F1_O*) = 320+77+62+54+28 = 541`. So 541 of the 3967 forecast values should have a censored value of 0 (i.e. raw value <60). I also turned on the MPR output line and ran this stat_analysis job to double-check. Here I'm just filtering the MPR lines, picking out the ones where the FCST column has a value of 0.
```
bin/stat_analysis -lookin out/point_stat/point_stat_360000L_20070331_120000V_mpr.txt -job filter -line_type MPR -column_thresh FCST eq0 -dump_row dump.txt
...
DEBUG 2: Processing 1 STAT files.
DEBUG 2: STAT Lines read     = 3967
DEBUG 2: STAT Lines retained = 541
```
The log messages confirm that 541 of the MPR lines have a FCST value of 0. And that is consistent with our understanding of the MCTC output.

I'm sure this could use an expanded explanation in [this section](https://met.readthedocs.io/en/latest/Users_Guide/point-stat.html?highlight=MCTC#measures-for-categorical-variables) of the MET User's Guide!

**Reply by malloryprow** [reply]
Ahhhh I see so with a treshold of ge1 it isn't inclusive of all values ge1 rather ge1 and lt2. Got it. Thank you!
